[AIEX] Add memset optimizations

andcarminati opened the pull request:
The idea is to reduce the number of stores related to memsets, avoiding whenever possible byte stores.

Current optimizations:

1. Memset align: for big stack objects, we align them to use vector stores.
2. Memset peeling: peel byte  and/or short stores out from memsets until we reach alignment, so we can avoid full byte scalarization of memsets.
3. Store combination: we combine stores into memsets to reach optimal store sequences. We can also take the leftovers from a previous memset expansion to create smaller vector stores.

andcarminati committed:
[AIEX] Align stack memsets to allow vector stores

In case of stack objects, if they are big enough, we
can improve alignment.

andcarminati committed:
[AIEX] Peel stores from memsets to reach alignment

We can peel some stores to reach alignment on memsets.
The alignment of the base pointer can be evaluated using
other uses of the pointer.

andcarminati committed:
[AIEX] Add base test for aligning memsets on stack objects

andcarminati committed:
[AIEX] Add base test for memset peeling

andcarminati commented on llvm/test/CodeGen/AIE/GlobalISel/combine-memset-stack-align.mir:
```diff
@@ -419,47 +315,215 @@ body:             |
-    ; LEGALIZED-NEXT: [[C24:%[0-9]+]]:_(s20) = G_CONSTANT i20 48
-    ; LEGALIZED-NEXT: [[PTR_ADD11:%[0-9]+]]:_(p0) = G_PTR_ADD [[FRAME_INDEX]], [[C24]](s20)
-    ; LEGALIZED-NEXT: G_STORE [[C1]](s32), [[PTR_ADD11]](p0) :: (store (s32) into %ir.object1 + 48, align 8)
+    ; LEGALIZED-NEXT: [[COPY7:%[0-9]+]]:_(s32) = COPY [[AIE_SEXT_EXTRACT_VECTOR_ELT]](s32)
```
Verify this!

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -13,6 +13,7 @@
 #include "AIE2TargetMachine.h"
 #include "AIEBaseInstrInfo.h"
 #include "MCTargetDesc/AIE2MCTargetDesc.h"
+#include "llvm/ADT/STLExtras.h"
```
Check includes.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3711,3 +3712,210 @@ bool llvm::matchPhiBitcast(MachineInstr &Phi, MachineRegisterInfo &MRI,
 
   return true;
 }
+bool llvm::matchAlignMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
```
nit: new line.

andcarminati replied:
Maybe force alignment here?

andcarminati commented:
TODO: add tests with different initializers.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3711,3 +3712,210 @@ bool llvm::matchPhiBitcast(MachineInstr &Phi, MachineRegisterInfo &MRI,
+      if ((Offset + 1) % 4) {
+        // Store the next two bytes to align to the next word boundary.
+        Register DataRegAdjustedToS16 =
+            B.buildConstant(LLT::scalar(16), (Initializer << 8) | 1).getReg(0);
```
Wrong: should be `(Initializer << 8)  |  Initializer `.

andcarminati commented on llvm/test/CodeGen/AIE/GlobalISel/combine-memset-peeling.mir:
```diff
@@ -0,0 +1,614 @@
+    ; CHECK-NEXT: [[C2:%[0-9]+]]:_(s20) = G_CONSTANT i20 45
+    ; CHECK-NEXT: [[PTR_ADD:%[0-9]+]]:_(p0) = G_PTR_ADD [[COPY]], [[C2]](s20)
+    ; CHECK-NEXT: G_STORE [[C1]](s8), [[PTR_ADD]](p0) :: (store (s8))
+    ; CHECK-NEXT: [[C3:%[0-9]+]]:_(s16) = G_CONSTANT i16 1
```
Wrong constant here!

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3711,3 +3712,210 @@ bool llvm::matchPhiBitcast(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  const TargetFrameLowering *TFI = MF->getSubtarget().getFrameLowering();
+  const int FrameIndex = DefDataInst->getOperand(1).getIndex();
+  int64_t BiggestAlignment = TFI->getStackAlignment();
+  const int64_t Offset = MFI.getObjectOffset(FrameIndex);
```
This information may not be accurate, can be zero!

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3711,3 +3712,210 @@ bool llvm::matchPhiBitcast(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  const int64_t Offset = MFI.getObjectOffset(FrameIndex);
+
+  // Convert to a power-of-two alignment.
+  while ((Offset % BiggestAlignment) != 0)
```
Change the approach to work directly on fixing the alignment on the stack object and memset.

andcarminati commented on llvm/test/CodeGen/AIE/GlobalISel/combine-memset-stack-align.mir:
```diff
@@ -0,0 +1,376 @@
+tracksRegLiveness: true
+stack:
+  - { id: 0, name: object0, type: default, offset: 0, size: 32,
+      alignment: 8, stack-id: default, callee-saved-register: '', callee-saved-restored: true,
```
Could we generate check lines for this alignment?

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3711,3 +3716,62 @@ bool llvm::matchPhiBitcast(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  const unsigned BasicVectorByteSize = TII.getBasicVecRegSize() / 8;
+
+  const Register CountReg = MI.getOperand(2).getReg();
+  auto Cst = getIConstantVRegValWithLookThrough(CountReg, MRI);
```
expand auto?

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3711,3 +3716,62 @@ bool llvm::matchPhiBitcast(MachineInstr &Phi, MachineRegisterInfo &MRI,
+    return false;
+
+  const Register PtrReg = MI.getOperand(0).getReg();
+  const MachineInstr *DefDataInst = MRI.getUniqueVRegDef(PtrReg);
```
const DefDataInst

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3775,3 +3775,167 @@ bool llvm::matchAlignMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+      if ((Offset + 1) % 4) {
+        // Store the next two bytes to align to the next word boundary.
+        Register DataRegAdjustedToS16 =
+            B.buildConstant(LLT::scalar(16), (Initializer << 8) | Initializer)
```
We've sign-extended Initializer. Could the sign bits  overflow into  the next byte  using this bitwise or, i.e. should we mask?

andcarminati replied:
Good catch!

andcarminati committed:
[AIEX] Add optimal Zero Store combiner

In the end, we try to reach an optimal number of stores
using the sinnergy with memset expander.

andcarminati committed:
[AIEX] Add a base test for store recombination

andcarminati commented:
QoR Results:

PM Size:
<img width="4400" height="1800" alt="Core_PMSize" src="https://github.com/user-attachments/assets/962ede70-6333-4896-90e5-61e2c6ac41b7" />

Note that important kernels were positively affected.
Avg. reduction: -1.68%

Instruction count:
<img width="4400" height="1800" alt="Core_Insn_Count" src="https://github.com/user-attachments/assets/0deda5a7-c398-4c2a-9691-a87a995270b9" />
Avg. reduction: -0.49%

Stack reduction:
<img width="4400" height="1800" alt="Core_StackSize" src="https://github.com/user-attachments/assets/b8c54cda-f733-475b-b90d-c0bde57ab268" />

The regressions are a side-effect of using one ne more vector. In case of function calls those vectors are saved. In absolute terms, we increase the stack in 64 bytes.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  };
+
+  return true;
+}
```
nit: new line.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  if (!RegAndOffset)
+    return false;
+
+  const unsigned Offset = RegAndOffset->second;
```
nit: signed type.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+    return false;
+
+  // We can merge over aligned small types, provided that
+  // the root ot the memset (first store) is a byte or short
```
nit: of

F-Stuckmann commented on llvm/test/CodeGen/AIE/GlobalISel/combine-seq-stores-memset.mir:
```diff
@@ -0,0 +1,343 @@
+...
+
+---
+name:            combine_to_s32_mixed
```
nice test

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  assert(MI.getOpcode() == TargetOpcode::G_MEMSET && "Expected a G_MEMSET");
+
+  if (!MemsetOptimizations)
+    return false;
```
nit: add debug message, that mem set optimizations are disabled and will be skipped

andcarminati replied:
As this is on by on by default, I see no point for this message.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+    return false;
+
+  // Try to keep alignment increase as minimum.
+  const unsigned BasicVectorByteSize = TII.getBasicVecRegSize() / 8;
```
In AIE2P this return 256, but we could also perform 128 bit stores, right?

andcarminati replied:
Good catch!

F-Stuckmann commented on llvm/test/CodeGen/AIE/GlobalISel/combine-memset-stack-align.mir:
```diff
@@ -0,0 +1,356 @@
+
+  define fastcc void @aligned_object_exact_size() personality ptr null {
+  entry:
+    %object0 = alloca [32 x i8], i32 0, align 8
```
Could you add a test with 16xi8 size (128 bits)

F-Stuckmann commented on llvm/test/CodeGen/AIE/GlobalISel/combine-seq-stores-memset.mir:
```diff
@@ -0,0 +1,343 @@
+# Negative test cases:
+
+---
+name:            no_combine_avoid_combiner_loop
```
CHECK: we can't combine here, because we aren't byte aligned?

F-Stuckmann commented on llvm/test/CodeGen/AIE/GlobalISel/combine-seq-stores-memset.mir:
```diff
@@ -0,0 +1,343 @@
+...
+
+---
+name:            no_combine_to_s32_load
```
CHECK: we can't combine, because of alias analysis in this case?

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+}
+
+static std::optional<std::pair<Register, int64_t>>
+getPtrAndConstantOffset(const MachineInstr *MI, unsigned PointerIndex,
```
I feel like this could be also useful in`AIEClusterBaseAddress.cpp`, maybe move this into a util header

andcarminati replied:
If we see the need, we can move for sure.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+}
+
+static std::optional<std::pair<Register, int64_t>>
+getPtrAndConstantOffsetFromReg(Register PtrReg, MachineRegisterInfo &MRI) {
```
why isn't this merged with `getPtrAndConstantOffset` ? I think these two functions could be combined.

andcarminati replied:
They have different use cases.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+static bool isBasePointerWordAligned(Register BasePtr,
+                                     MachineRegisterInfo &MRI) {
+
+  auto IsBasePointerAligned = [&](const MachineInstr *MI, int64_t Offset) {
```
nit: This name could be updated, since we only care for stores.

andcarminati replied:
We care about loads and stores.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+                                     MachineRegisterInfo &MRI) {
+
+  auto IsBasePointerAligned = [&](const MachineInstr *MI, int64_t Offset) {
+    if (!isa<GLoadStore>(MI))
```
If you extend this check to not-exit with `G_MEMSET` you could reuse this function in `matchAlignMemset`

andcarminati replied:
The purpose of this function is just to prove that we have at least one aligned load/store related to the base pointer, we simply don't care about that the load or store is used for. We simply look to the others uses of the same pointer. For the matchAlignMemset we to the target instruction directly.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+      auto RegAndOffset = getPtrAndConstantOffset(&MI, 0, MRI);
+      if (!RegAndOffset)
+        continue;
+      if (any_of(MRI.use_instructions(RegAndOffset->first),
```
what happens if a store is basePointerAligned and another is not basePointerAligned?
I don't think the method should return in that case

andcarminati replied:
I one instruction is aligned, that is all we need.

martien-de-jong replied:
Perhaps add a comment. The actual value must have the best alignment of all its uses.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+                   return IsBasePointerAligned(&UseMI, RegAndOffset->second);
+                 }))
+        return true;
+    } else if (IsBasePointerAligned(&MI, 0)) {
```
Analogues here, wouldn't we want all stores to be base pointer aligned, not just the first?

andcarminati replied:
If we see one aligned operation, it is safe to state the the pointer is aligned, otherwise we have wrong code from the beginning. Unaligned operations can be legalized by the legalizer, but wrong align information is uncoverable, so we can trust.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+
+  // If it is already aligned we have nothing to do.
+  if (MMOAlign.value() % 4 == 0)
+    return false;
```
You could reuse an updated version of `IsBasePointerAligned` here ;)

andcarminati replied:
Here we care about one specific `memset`, `IsBasePointerAligned` is just to prove that we have an aligned operation anywhere else, that allow us to enable this current optimization.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  const auto Cst = getIConstantVRegValWithLookThrough(SizeReg, MRI);
+  if (!Cst)
+    return false;
+  const int64_t Size = Cst->Value.getSExtValue();
```
With a more generic version of `getPtrAndConstantOffset`, we could also use it here.

andcarminati replied:
This is not related to the pointer, this is the size register of the memset, which represents the number o bytes to be initialized.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  auto Cst = getIConstantVRegValWithLookThrough(CountReg, MRI);
+  if (!Cst)
+    return false;
+  const unsigned ByteCount = Cst->Value.getZExtValue();
```
With a more generic version of getPtrAndConstantOffset, we could also use it here.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+    return false;
+
+  Register PtrReg = MI.getOperand(0).getReg();
+  const Register DataReg = MI.getOperand(1).getReg();
```
This could also use a generic variant of `getPtrAndConstantOffset`

andcarminati replied:
This is the data register to be written, offset and base pointers are retrieved below.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+    auto &MF = B.getMF();
+    unsigned OutlinedStores = 0;
+
+    auto BuildPADD = [&](int64_t CurrentOffset) {
```
nit: we are not using `CurrentOffset`

andcarminati replied:
This was great catch!

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+      }
+
+      // Dont't peel two more if we are aligned.
+      if ((Offset + 1) % 4) {
```
CHECK: we enter this function, even if we already aligned the offset to a short? Don't we have to update the Offset in that case? 
Also in the case of Short alignment, isn't this check wrong? 

Also: I think we could reduce the nesting here for clarity, if we used something like `CurrentOffset`
`if (CurrerntOffset % 4 ==0 ) // skip peeling`
``
if (CurrentOffset %2) // handle byte allignment and update CurrentOffset
``
`if (CurrentOffset % 4) // handle short alignment `

andcarminati replied:
Note that this function will always update the alignment of the memset to 4, if we can prove that the pointer is proper aligned, so next time we will skip because we test this at very beginning. If we just update alignment, Offset should be the same.

We cannot strip out the first case to be independent because we could have a realignment opportunity. The next version of this PR will include a more verbose alignment check.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+      }
+    }
+
+    const unsigned NewSize = Size - OutlinedStores;
```
nit rename `OutlineStores` to `PeeledBytes`, since this is what you actually did and makes this arithmetic here clearer.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+    const int64_t MemsetOffset = Offset + OutlinedStores;
+    assert(MemsetOffset % 4 == 0 && "Memset still unaligned?");
+    // Now, what remains is aligned, we just need to fix Offset, Size and MMO.
+    MachineMemOperand *NewMMOMemSet = BuildMMO(MMO->getSize(), Align(4));
```
CHECK: if we already peeled stores, shouldn't the new MMO have a smaller size than the original one?

andcarminati replied:
This is related to the location size:
```
G_MEMSET %0(p0), %1(s8), %3(s20), 0 :: (store (s8) into %ir.object0, align 1)
```

Which is always `(s8)` for a memset.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+    MI.dropMemRefs(MF); // Safe to drop the MMO now.
+    MI.addMemOperand(MF, NewMMOMemSet);
+    MI.getOperand(2).setReg(
+        B.buildConstant(LLT::scalar(20), NewSize).getReg(0));
```
Is the MMO still up to date, if the we use NewSize here?

andcarminati replied:
Yes, see my previous comment about LocationSize.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+}
+
+static std::optional<std::pair<Register, int64_t>>
+getPtrAndConstantOffsetFromStore(const GStore *StMI, MachineRegisterInfo &MRI) {
```
nit: I think a generalized `getPtrAndConstantOffset` could also handle this case

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+
+  // Small alignments are not interesting, we are trying to match
+  // vector stores here.
+  const bool IsVectorAlignment = MMOAlign.value() >= 16;
```
nit: move this Definition to where it is first used. Up here it adds confusion with the comment.

andcarminati replied:
I improved the comment.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+    if (DataType.getSizeInBytes() >= MMOAlign.value())
+      return 0;
+    auto Cst = getIConstantVRegValWithLookThrough(DataReg, MRI);
+    return (Cst && Cst->Value.getZExtValue() == 0) ? DataType.getSizeInBytes()
```
Why are we limited to 0 here? I guess, this would still work with any constant, right?

andcarminati replied:
Discussed offline, this leads to impractical cases.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  const LLT DataType = MRI.getType(DataReg);
+
+  // We already have a vector store.
+  if (DataType.getSizeInBytes() >= MMOAlign.value())
```
I don't think this check here is necessary, since we also check it in line 4099

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+                                                   : 0;
+  };
+
+  auto PtrAndOffset = getPtrAndConstantOffsetFromStore(&StMI, MRI);
```
nit: For clarity I would move this right before the first use

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  //   STORE i8 0, [p0+3]
+  // Will be transformed to:
+  //   STORE i32 0, [p0] (Align 16)
+  if ((IsVectorAlignment && NumberOfBytes < 16) || MatchedSeqStores.empty())
```
I feel like the minimum number of bytes should be something from TTA like `getMinVectorBitSize`.

andcarminati replied:
Yeah, currently it matches all the targets, but in the future we may need some change.

F-Stuckmann commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3783,3 +3791,395 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  // This will cause a loop. We prevent by restricting
+  // combinations that will expand again to the same
+  // types: 3 bytes.
+  if (NumberOfBytes == 3)
```
Would we also have the same issue with `NumberOfBytes == 2`? Who is rotating the stores ?

andcarminati replied:
No. This happens only with 3 bytes:

```
G_STORE s8
G_STORE s16

then:
G_STORE s16
G_STORE s8

then:
G_STORE s8
G_STORE s16

....
```

If we have two byte stores, they will be merged into a short store, and that is all.

martien-de-jong replied:
I understand it, but I don't like it. The behaviour of the memset expander is encrypted here in a magic literal '3'. Also, I think we will not coalesce three byte stores under this condition?

andcarminati replied:
Indeed I detest magic numbers, but we have a reason and a test for it. As far I remember, I tripped with this case in a real benchmark.

F-Stuckmann commented on llvm/test/CodeGen/AIE/GlobalISel/combine-memset-peeling.mir:
```diff
@@ -0,0 +1,722 @@
+    %1:_(s32) = G_CONSTANT i32 0
+    %2:_(s8) = G_CONSTANT i8 0
+    %3:_(s32) = G_CONSTANT i32 7
+    %5:_(s20) = G_CONSTANT i20 45
```
Could we prove the alignment, if the offset here would maintain the alignment?

F-Stuckmann replied:
We do know the alignment of the memset here. Where is the difference between this and `peel_3_bytes`

andcarminati replied:
This could be a further improvement if we found one case. In this case, we could compare the two offsets to see if the difference is aligned...

andcarminati replied:
The difference is the uncommon odd offset.

F-Stuckmann commented on llvm/test/CodeGen/AIE/GlobalISel/combine-memset-peeling.mir:
```diff
@@ -0,0 +1,722 @@
+    %3:_(s32) = G_CONSTANT i32 7
+    G_STORE %1(s32), %0(p0) :: (store (s32), align 2)
+    %4:_(s20) = G_TRUNC %3(s32)
+    G_MEMSET %0(p0), %2(s8), %4(s20), 1 :: (store (s8))
```
Wouldn't this memset be at least aligned by 2, since it uses the same base pointer as the aligned store?

andcarminati replied:
The current implementation only considers word alignment. It could be a future improvement.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3787,3 +3792,64 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  if (DefDataInst->getOpcode() != TargetOpcode::G_FRAME_INDEX)
+    return false;
+
+  MachineMemOperand *MMO = MI.memoperands().front();
```
Can this fail if there are no memoperands?

andcarminati replied:
You are right, I will insert a check.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3787,3 +3792,64 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  if (MMOAlign == OptimalAlign)
+    return false;
+
+  MatchInfo = [=, &MI, &MFI](MachineIRBuilder &B) {
```
nit: no need to capture MF or MFI. We can fetch them here from MI

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3787,3 +3792,64 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+    return false;
+
+  MatchInfo = [=, &MI, &MFI](MachineIRBuilder &B) {
+    MFI.setObjectAlignment(FrameIndex, OptimalAlign);
```
I'm on unfamiliar grounds here, but should we check that this hasn't influenced some stack layout already?

andcarminati replied:
The complete stack layout is finalized very late in the pipeline, basically it is safe to change. Some other combiners can also increase the alignment similarly, like the memset expander.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3787,3 +3792,64 @@ bool llvm::matchPhiOfUndef(MachineInstr &Phi, MachineRegisterInfo &MRI,
+  MachineFrameInfo &MFI = MF->getFrameInfo();
+  const int FrameIndex = DefDataInst->getOperand(1).getIndex();
+
+  const Align OptimalAlign = Align(BasicVectorByteSize);
```
nit: if we only use half vectors, could we use halfvector alignment to save on alignment gaps?

andcarminati replied:
I think it is better if we can choose the alignment based on `ByteCount`.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3853,3 +3854,176 @@ bool llvm::matchAlignMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+      return false;
+    const MachineMemOperand *MMO = MI->memoperands().front();
+    return MMO && matchAlignment<4>(MMO->getAlign().value()) &&
+           matchAlignment<4>(Offset);
```
I'm not sure whether we can always use front(), since that would fail on an empty memoperand range? And then, if we do have memoperands, can it return a nullptr, i.e. is the check for non-null MMO necessary?

martien-de-jong replied:
Perhaps wrap in a getMemOperandOrNull() utility?

andcarminati replied:
I will include a test to see if it is empty.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3853,3 +3854,176 @@ bool llvm::matchAlignMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+  // If we are short-aligned we can generate
+  //  - G_STORE s16
+  //  - G_MEMSET (... n - 2 ...)
+  if (Size < 3)
```
Does it matter to bail out here? we are doing the alignment case analysis anyway. If we reach zero size by peeling  we remove the memset and we're done.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3853,3 +3854,176 @@ bool llvm::matchAlignMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+}
+
+template <uint64_t TargetAlign> constexpr bool matchAlignment(uint64_t Value) {
+  return isAligned(Align(TargetAlign), Value);
```
I find this a clumsy interface. I think the caller could usefully employ Value % TargetAlign, which should be brought down to zero, That could be done in a loop processing successive powers of two. The initial value is the shift in offset that you need to adjust the memset.

andcarminati replied:
Maybe it is not great, I tried to give a name to the alignment check and reuse `isAligned` function. About the rest of the implementation, but I don't thing that turning the caller code into a loop will help because, to align a memset, we should peel between one and two stores only, and each store is different (different MMO, type and constant to store).

martien-de-jong replied:
I grant you the loop argument. But just have case 0, 1, 2, and 3 in a switch rather than repeated calls to matchAlignment would lead to simpler control flow. bottom line is that we don't want to know whether it's aligned, but in what way it is unaligned.

andcarminati replied:
I would like to stick to the current approach, because I feel it simple and correct. Let's say, if we are not aligned, we just align on the fly base on the current offset. I also update that code with a fix and some comments.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3853,3 +3854,176 @@ bool llvm::matchAlignMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+        continue;
+      if (any_of(MRI.use_instructions(RegAndOffset->first),
+                 [&](const MachineInstr &UseMI) {
+                   return IsBasePointerAligned(&UseMI, RegAndOffset->second);
```
nit: the Offset part of this is invariant, and can be tested outside of the any_of()

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -4027,3 +4036,168 @@ bool llvm::matchPeelMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+  return getPtrAndConstantOffsetFromReg(StMI->getPointerReg(), MRI);
+}
+
+// To make a store dead, we convert it to a dead ADDs and let DCE to do the
```
nit: a dead G_ADD

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -4027,3 +4036,168 @@ bool llvm::matchPeelMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+  StMI->setDesc(TII.get(TargetOpcode::G_ADD));
+  for (unsigned I = 0; I < 3; I++) {
+    StMI->addOperand(MachineOperand::CreateReg(
+        I == 0 ? MRI.cloneVirtualRegister(DataReg) : DataReg, I == 0));
```
nit: just an addDef and two addUses without further conditionals?

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -4027,3 +4036,168 @@ bool llvm::matchPeelMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+
+  // Small alignments are less interesting, we are trying to match
+  // vector stores here. Even though we can match some byte/short
+  // stores to to word stores.
```
nit: to to

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -4027,3 +4036,168 @@ bool llvm::matchPeelMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+  // Small alignments are less interesting, we are trying to match
+  // vector stores here. Even though we can match some byte/short
+  // stores to to word stores.
+  const bool IsVectorAlignment = MMOAlign.value() >= MinVectorStoreSize;
```
declare closer to its first use?

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -4027,3 +4036,168 @@ bool llvm::matchPeelMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+  if (!IsVectorAlignment && DataType.getSizeInBytes() >= 4)
+    return false;
+
+  // Are we storing Zero? if yes, how many? Zero bytes means not a Zero store at
```
Overloading Zero here. ...A count of zero means....

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -4027,3 +4036,168 @@ bool llvm::matchPeelMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+    if (DataType.getSizeInBytes() >= CurrMMOAlign.value())
+      return 0;
+    auto Cst = getIConstantVRegValWithLookThrough(DataReg, MRI);
+    return (Cst && Cst->Value.getZExtValue() == 0) ? DataType.getSizeInBytes()
```
There's APInt::IsZero()

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -4027,3 +4036,168 @@ bool llvm::matchPeelMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+
+      auto PtrAndOffset = getPtrAndConstantOffsetFromStore(CurrSt, MRI);
+      if (!PtrAndOffset) // Non-constant offset.
+        return false;
```
Aren't we wasting the accumulated MatchedSeqStores here?

andcarminati replied:
Good catch!

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -4027,3 +4036,168 @@ bool llvm::matchPeelMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+  // Will be transformed to:
+  //   STORE i32 0, [p0] (Align 16)
+  if ((IsVectorAlignment && NumberOfBytes < MinVectorStoreSize) ||
+      MatchedSeqStores.empty())
```
nit: I think the empty() test should be hoisted to right after the collection loop. That separates the 'no opportunities found' from the 'not profitable' criteria.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -4027,3 +4036,168 @@ bool llvm::matchPeelMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
+
+      // Pointers are different or we have non-linear store.
+      if ((CurrPtr != Ptr) || (ExpectedOffset != CurrOffset))
+        break;
```
Would it be difficult to track a bitvector with written zero bytes on the same pointer so long as there's no overlap or other dangers and check for a consecutive range afterwards? That would take care of the strict ordering we now need.

andcarminati replied:
I would say that is this is feasible. We could have a follow-up work that handles gaps and non-zero values if we see practical cases.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -4160,8 +4161,7 @@ bool llvm::matchSequentialStores(GStore &StMI, MachineRegisterInfo &MRI,
   //   STORE i32 0, [p0] (Align 16)
-  if ((IsVectorAlignment && NumberOfBytes < MinVectorStoreSize) ||
-      MatchedSeqStores.empty())
+  if ((IsVectorAlignment && NumberOfBytes < MinVectorStoreSize))
```
nit: We can drop a pair of parentheses now.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3995,36 +3994,38 @@ bool llvm::matchPeelMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
       return NewMMO;
     };
 
+    int64_t PeelOffset = Offset;
```
Yes! PeelOffset represents an invariant of the process that fixes the alignment.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3995,36 +3994,38 @@ bool llvm::matchPeelMemset(MachineInstr &MI, MachineRegisterInfo &MRI,
     }
 
-    const int64_t MemsetOffset = Offset + PeeledStores;
+    const int64_t MemsetOffset = PeelOffset;
```
Or even copy-propagate it to its use.

martien-de-jong reviewed (dismissed):
Good to go.
